lesson20/index.js

```diff
@@ -0,0 +1,130 @@
+const parent = document.querySelector("body");
+
+const createElementWithClassName = (type, className) => {
+    const element = document.createElement(type);
+    element.className = className;
+    return element;
+};
+
+const addLoading = () => {
+    const img = document.createElement("img");
+    const imgWrapper = createElementWithClassName("div", "loading");
+
+    img.src = "loading-circle.gif";
+    imgWrapper.id = "js-loading";
+    parent.appendChild(imgWrapper).appendChild(img);
+};
+
+const removeLoading = () => {
+    document.getElementById("js-loading").remove();
+};
+
+const fetchData = async(endpoint) => {
+    const response = await fetch(endpoint);
+
+    if(!response.ok){
+        const errorMessage = `${response.status}:${response.statusText}`;
+        parent.appendChild(createErrorMessage(errorMessage));
+        console.error(errorMessage);
+        return;
+    }
+    return await response.json();
+};
+
+const fetchUserData = async() => {
+    return new Promise((resolve) => {
+        setTimeout(() => resolve(fetchData("https://mocki.io/v1/8853f6f1-f1f0-4db3-9788-668933e2f788")), 3000);
+        // setTimeout(() => resolve(fetchData("https://mocki.io/v1/f12ae2d1-310d-4120-8749-47773d65e236")), 3000);//空配列
+        // setTimeout(() => resolve(fetchData("https://httpstat.us/503")), 3000);//503 error
+    })
+};
+
+const init = async() => {
+    addLoading();
+    try {
+        const json = await fetchUserData();
+        if (!json) return;
+
+        const data = json.data;
+        renderTable(data);
+
+        if (data.length === 0) {
+            parent.textContent = "まだデータがありません";
+            console.log("まだデータがありません");
+        }
+        return data;
+
+    } catch(e) {
+        console.error(e);
+        parent.appendChild(createErrorMessage(e));
+    } finally {
+        removeLoading();
+    }
+};
```

Comment:
IMO
 I don't think renderTable() should be in a try. This is because we cannot determine if the error is due to "promote" or "renderTable()". (I received this review when I was challenging Task10.)
I will suggest like below.

```suggestion
const getData = async() => {
    addLoading();
    try {
        const json = await fetchUserData();
        if (!json) return;

        const data = json.data;
        renderTable(data);

        if (data.length === 0) {
            parent.textContent = "まだデータがありません";
            console.log("まだデータがありません");
        }
        return data;

    } catch(e) {
        console.error(e);
        parent.appendChild(createErrorMessage(e));
    } finally {
        removeLoading();
    }
};

const init = async () => {
  const data = await getData();
  data && renderTable(data);
};
```

Reply:
Thank you for your review!!

> we cannot determine if the error is due to "promise" or "renderTable()"

I thought that was right. I was very shallow in my thinking...! 
I am not good at try,catch, so your advice helped me to understand better.
Thank you so much!

I fixed it.
[34ebdb3](https://github.com/haru-programming/JavaScript-lessons/pull/25/commits/34ebdb36bdd11d1e5163003c8b18798f5332358f)